cpp/src/arrow/array/builder_nested.cc

```diff
@@ -110,12 +108,25 @@ Status MapBuilder::AppendNull() {
 
 Status MapBuilder::AppendNulls(int64_t length) {
   DCHECK_EQ(item_builder_->length(), key_builder_->length());
+  RETURN_NOT_OK(CheckStructBuilderLength());
   RETURN_NOT_OK(list_builder_->AppendNulls(length));
   length_ = list_builder_->length();
   null_count_ = list_builder_->null_count();
   return Status::OK();
 }
 
+Status MapBuilder::CheckStructBuilderLength() {
+  // If key/item builders have been appended, adjust struct builder lenght
+  // to match. Struct and key are non-nullable, append all valid values.
+  if (list_builder_->value_builder()->length() < key_builder_->length()) {
```

Comment:
It's comparing `list_builder_->value_builder()` which is the same as `struct_builder`. I just do the `check_cast` in the `if` statement in case it is not needed. I could move the `checked_cast` to the top though, if it would be more clear?

Reply:
Yes, it would be more clear IMHO. Also, the `checked_cast` is almost a no-op in release mode.